File: code/329/solution.cpp
#include "../template/prime.cpp"
#include <iostream>/*{{{*/
#include <map>
#include <math.h>
#include <set>
#include <vector>
const long long e2 = 100;
const long long e3 = e2 * 10;
const long long e4 = e3 * 10;
const long long e5 = e4 * 10;
const long long e6 = e4 * e2;
const long long e7 = e6 * 10;
const long long e8 = e7 * 10;
const long long e9 = e8 * 10;
long long _mod     = e9 + 7;
using namespace std; /*}}}*/
__int128 gcd(__int128 a, __int128 b) {
    if (a == 0)
        return b;
    return gcd(b % a, a);
}
struct fac {
    __int128 upper;
    __int128 lower;

    fac() {}
    fac(__int128 a, __int128 b) {
        __int128 g = gcd(a, b);
        upper      = a / g;
        lower      = b / g;
    }
    void show() { cout << (long long)upper << ' ' << (long long)lower << endl; }
};

fac mul(fac a, fac b) { return fac(a.upper * b.upper, a.lower * b.lower); }

fac add(fac a, fac b) { return fac(a.upper * b.lower + b.upper * a.lower, a.lower * b.lower); }

fac now[501];
void handle(bool prime) {
    fac nxt[501];
    for (int i = 1; i <= 500; i++)
        nxt[i] = fac(0, 1);
    nxt[2]   = mul(now[1], fac(prime ? 2 : 1, 3));
    nxt[499] = mul(now[500], fac(prime ? 2 : 1, 3));

    for (int i = 2; i < 500; i++) {
        fac buf(Prime::is_prime(i - 1) == prime ? 2 : 1, 6);
        nxt[i - 1] = add(nxt[i - 1], mul(now[i], buf));
        buf        = fac(Prime::is_prime(i + 1) == prime ? 2 : 1, 6);
        nxt[i + 1] = add(nxt[i + 1], mul(now[i], buf));
    }
    for (int i = 1; i <= 500; i++)
        now[i] = nxt[i];
}
int main() {
    Prime::init(600);
    for (int i = 1; i <= 500; i++) {
        now[i] = fac(Prime::is_prime(i) ? 2 : 1, 1500);
    }

    string query = "PPPPNNPPPNPPNPN";
    for (int i = 1; i < query.length(); i++) {
        handle(query[i] == 'P');
    }

    fac ans(0, 1);
    for (int i = 1; i <= 500; i++) {
        ans = add(ans, now[i]);
    }
    cout << (long long)ans.upper << ' ' << (long long)ans.lower << endl;
    return 0;
}

File: code/329/solution.py
from random import random
from math import *


def gcd(a, b):
    if a == 0:
        return b
    return gcd(b % a, a)


def format(a, b):
    g = gcd(a, b)
    return [a // g, b // g]


def mul(a, b):
    return format(a[0] * b[0], a[1] * b[1])


def add(a, b):
    return format(a[0] * b[1] + b[0] * a[1], a[1] * b[1])


is_prime = {}
now = [[0, 1] for i in range(501)]


def handle(p):
    global now
    nxt = [[0, 1] for i in range(501)]
    x = 1
    if p:
        x = 2
    nxt[2] = mul(now[1], [x, 3])
    nxt[499] = mul(now[500], [x, 3])

    for i in range(2, 500):
        buf = [1, 6]
        if is_prime[i - 1] == p:
            buf = [1, 3]
        nxt[i - 1] = add(nxt[i - 1], mul(now[i], buf))
        buf = [1, 6]
        if is_prime[i + 1] == p:
            buf = [1, 3]
        nxt[i + 1] = add(nxt[i + 1], mul(now[i], buf))
    for i in range(1, 501):
        now[i] = nxt[i]


if __name__ == "__main__":
    for i in range(1, 501):
        is_prime[i] = True
    for i in range(2, 501):
        if is_prime[i]:
            for j in range(i * 2, 501, i):
                is_prime[j] = False
    is_prime[1] = False
    for i in range(1, 501):
        now[i] = [1, 1500]
        if is_prime[i]:
            now[i] = [1, 750]
    query = "PPPPNNPPPNPPNPN"
    for i in range(1, len(query)):
        handle(query[i] == "P")
    ans = [0, 1]
    for i in range(1, 501):
        ans = add(ans, now[i])
        print(i, now[i])
    print(ans)
